evap/static/js/datagrid.js

```diff
@@ -0,0 +1,284 @@
+        this.storageKey = storageKey;
+        this.searchInput = searchInput;
+    }
+
+    attachTo(container) {
+        this.container = container;
+        this.columns = Array.from(this.headers.keys());
+        this.rows = this.fetchRowData();
+        this.loadFromStorage();
+        this.bindEvents();
+    }
+
+    bindEvents() {
+        this.delayTimer = null;
+        this.searchInput.on("change paste input", () => {
+            clearTimeout(this.delayTimer);
+            this.delayTimer = setTimeout(() => {
+                this.search(this.searchInput.val());
+            }, 200);
+        }).keypress(event => {
+            if (event.keyCode === 13) {
+                event.target.blur();
+            }
+        });
+
+        for (const [column, header] of this.headers) {
+            header.click(() => {
+                const ordering = header.hasClass("col-sorting-asc") ? "desc" : "asc";
+                this.order([[column, ordering]]);
+            });
+        }
+    }
+
+    fetchRowData() {
+        return this.container.children().get()
+            .map(row => {
+                let data = {
+                    element: row,
+                    order: new Map(),
+                    filter: new Map(),
+                };
+                data.search = this.findSearchableCells($(row))
+                    .flatMap(element => this.searchWordsOf($(element).text()));
+
+                this.addFilterRowData($(row), data);
+
+                for (const column of this.columns) {
+                    const cell = this.findCellByColumn($(row), column);
+                    if (cell.is("[data-order]")) {
+                        data.order.set(column, cell.data("order"));
+                    } else {
+                        data.order.set(column, cell.html().trim());
+                    }
+                }
+                return data;
+            });
+    }
+
+    searchWordsOf(string) {
+        return string.toLowerCase().trim().split(/\s+/);
+    }
+
+    search(searchValue) {
+        this.state.search = searchValue;
+        this.applyFilter();
+        this.update();
+    }
+
+    filter(filter) {
+        Object.assign(this.state.filter, filter);
+        this.applyFilter();
+        this.update();
+    }
+
+    applyFilter() {
+        const searchWords = this.searchWordsOf(this.state.search);
+        for (const row of this.rows) {
+            const isDisplayedBySearch = searchWords.every(searchWord => {
+                return row.search.some(rowWord => rowWord.includes(searchWord));
+            });
+            const isDisplayedByFilters = Object.entries(this.state.filter).every(([name, filterValues]) => {
+                const matchesAtLeastOneValue = filterValues.some(filterValue => {
+                    return row.filter.get(name).some(rowValue => rowValue === filterValue);
+                });
+                return filterValues.length === 0 || matchesAtLeastOneValue;
```

Comment:
I'd guess that the JS engines are not that smart and would check `filterValues.length === 0` before computing `matchesAtLeastOneValue`.

Reply:
Do you ask for performance considerations?
The reason why it is written like that: `some` will be `false` on an empty collection – but if nothing is selected in a filter category, we want to show everything

Reply:
I think I've understood what the code does and i think it is correct. Yes, I was purely looking at performance. Reordering such boolean expressions is what i did in 433dcf28b and it made a rather large difference.

To avoid misunderstandings, I propose to early-out on `filterValues.length === 0` before executing the `some` call. whether that's by inlining `matchesAtLeastOneValue` or adding an additional `if` i don't care. I also don't care if you show that all of this doesn't make any practical difference :)